geode-core/src/test/java/org/apache/geode/test/dunit/cache/rules/CacheRule.java

```diff
@@ -0,0 +1,185 @@
+import org.apache.geode.test.dunit.Disconnect;
+import org.apache.geode.test.dunit.VM;
+import org.apache.geode.test.dunit.rules.DistributedExternalResource;
+
+@SuppressWarnings({"serial", "unused"})
+public class CacheRule extends DistributedExternalResource {
+
+  private static InternalCache cache;
+
+  private final boolean createCacheInAll;
+  private final boolean createCache;
+  private final boolean disconnectAfter;
+  private final List<VM> createCacheInVMs;
+  private final Properties config;
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  CacheRule(final Builder builder) {
+    this.createCacheInAll = builder.createCacheInAll;
+    this.createCache = builder.createCache;
+    this.disconnectAfter = builder.disconnectAfter;
+    this.createCacheInVMs = builder.createCacheInVMs;
+    this.config = builder.config;
+  }
+
+  @Override
+  protected void before() {
+    if (createCacheInAll) {
+      invoker().invokeInEveryVM(() -> createCache(config));
+    } else {
+      if (createCache) {
+        createCache(config);
+      }
+      for (VM vm : createCacheInVMs) {
+        vm.invoke(() -> createCache(config));
+      }
+    }
+  }
+
+  @Override
+  protected void after() {
+    closeAndNullCache();
+    invoker().invokeInEveryVM(() -> closeAndNullCache());
+
+    if (disconnectAfter) {
+      Disconnect.disconnectAllFromDS();
+    }
+  }
+
+  public InternalCache getCache() {
+    return cache;
+  }
+
+  public InternalDistributedSystem getSystem() {
+    return cache.getInternalDistributedSystem();
+  }
+
+  public static InternalCache getCacheForTesting() {
+    return cache;
+  }
+
+  private static void createCache(final Properties config) {
+    cache = (InternalCache) new CacheFactory(config).create();
+  }
+
+  private static void closeAndNullCache() {
+    closeCache();
+    nullCache();
+  }
+
+  private static void closeCache() {
+    try {
+      if (cache != null) {
+        cache.close();
+      }
+    } catch (Exception ignored) {
+      // ignored
+    }
+  }
+
+  private static void nullCache() {
+    cache = null;
+  }
+
+  /**
+   * Builds an instance of SharedCountersRule
+   */
+  public static class Builder {
+
+    private boolean createCacheInAll;
+    private boolean createCache;
+    private boolean disconnectAfter;
+    private List<VM> createCacheInVMs = Collections.emptyList();
+    private Properties config = new Properties();
+
+    public Builder() {
+      config.setProperty(LOCATORS, getLocators());
+    }
+
+    /**
+     * Create Cache in every VM including controller and all DUnit VMs
+     */
+    public Builder createCacheInAll() {
+      createCacheInAll = true;
+      return this;
+    }
+
+    /**
+     * Create Cache in specified VM
+     */
+    public Builder createCacheIn(final VM vm) {
+      if (createCacheInVMs.isEmpty()) {
+        createCacheInVMs = new ArrayList<>();
+      }
+      if (!createCacheInVMs.contains(vm)) {
```

Comment:
Should createCacheInVMs be a set instead (since you are handling it that way)? That would influence the parent class as well, but it looks like there too duplicates would be bad.

Reply:
I wanted createCacheInVMs to be ordered by add-order. I couldn't use a SortedSet because VM does not implement Comparable and even then the order would be wrong. Another bood alternative would be to implement Comparable to order by getId().

Reply:
LinkedHashSet should do the trick for insertion-order iteration.